### ISSUE TITLE
ICMP6 Input Handlers - enum with return codes instead of void and gotos

### ISSUE THREAD

**Issue by joakimeriksson** [contiki-os/contiki#1237]
I have taken a look at the ICMP6 handlers and they introduced a bunch of "bugs" and retransmission of input packets which is now fixed by a large number of goto's which is probably all ok but I think a bit risky since it is easy to do a return; and forget that we need to drop.

I would be happy to fix this with a return value enum so that all handler in each place there is a return need to decide on dropping or sending in a clear way.

Something like:
```
typedef enum {
  UIP_ICMP6_INPUT_OP_SEND,
  UIP_ICMP6_INPUT_OP_DROP,
  UIP_ICMP6_INPUT_ERROR,
  UIP_ICMP6_INPUT_NO_HANDLER
} uip_icmp6_input_handler_ret_t;
```

And a switch for the return value in the uip6 code that calls the handler code +
some minor changes to the handler call code.

Then at each goto discard it will be a return UIP_ICMP6_INPUT_OP_DROP;
and for each return; it will be a return UIP_ICMP6_INPUT_OP_SEND;

Then it is much harder to do mistakes when adding new code since you will need to take a decision on what to return.

**Comment by laurentderu** [comment]
Indeed, the actual behaviour of ICMP handlers is too risky, at least this solution will force to choose a behaviour at compilation time, so I'm clearly in favour of this one.
